### PR TITLE
Backport the fixes related to replacing ApacheDS with UnboundID from pull #1783 to the master #1846 And Upgrade our UnboundID dependencies from 4.0.9 to 5.1.1 #1858 #105

### LINKED ISSUE #1846: Backport the fixes related to replacing ApacheDS with UnboundID from pull #1783 to the master
# Task Description

We should backport the fixes from #1783 that related to ApacheDS being replaced with UnboundID to the `master`. The version of the `com.unboundid:unboundid-ldapsdk:5.0.1` (see [here](https://github.com/strongbox/strongbox/pull/1783/files#diff-f96b8068a631d962b8fd30f5eaaedb24R164)) should be extracted to the [strongbox-parent](https://github.com/strongbox/strongbox-parent/) project (you will find [this article](https://strongbox.github.io/developer-guide/upgrading-dependencies-and-plugins.html) useful, as it relates to adding/updating dependencies in the `strongbox-parent`).

# Tasks

The following tasks will need to be carried out:
* [ ] Define the `com.unboundid:unboundid-ldapsdk:5.0.1` dependency in the `<dependencyManagement/>` of the [strongbox-parent](https://github.com/strongbox/strongbox-parent/).
* [ ] Backport only the necessary sub-set of changes from #1783 (including the [LDIF fixes](https://github.com/strongbox/strongbox/pull/1783/files#diff-cd5cfc21b902bf2e72d0aee2b128ed40R19-R62)).
* [ ] Test that things still work properly.

# Task Relationships

This task:
* Relates to: #1783 

# Useful Links

* [Upgrading Dependencies And Plugins](https://strongbox.github.io/developer-guide/upgrading-dependencies-and-plugins.html)

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @steve-todorov

### PR DESCRIPTION
**Pull request by mknutsen79**
# Pull Request Description

This pull request fixes #1846 .

# Acceptance Test

* [x] Building the code with `mvn clean install -Dintegration.tests` still works.
* [x] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [x] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [x] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [x] Yes, please see strongbox/strongbox-parent#113
  * [ ] No

* Does this require an update of the documentation?
  * [ ] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [x] No

# Code Review And Pre-Merge Checklist

* [ ] My code follows the [coding convention](https://strongbox.github.io/developer-guide/coding-convention.html) of this project.
* [x] I have performed a self-review of my own code.
* [ ] I have commented my code in hard-to-understand areas.
* [x] My changes generate no new warnings.
* [ ] I have added tests that prove my fix is effective or that my feature works.
* [x] New and existing unit tests pass locally with my changes.

### CODE REVIEW

File: strongbox-cron/strongbox-cron-tasks/pom.xml [hunk 1]
```diff
@@ -163,12 +163,10 @@
...
-            <groupId>org.apache.directory.server</groupId>
-            <artifactId>apacheds-core</artifactId>
-        </dependency>
-        <dependency>
-            <groupId>org.apache.directory.server</groupId>
-            <artifactId>apacheds-protocol-ldap</artifactId>
+            <groupId>com.unboundid</groupId>
+			<artifactId>unboundid-ldapsdk</artifactId>
+			<version>5.1.1</version>
+			<scope>test</scope>
```
Comment (carlspring) [inline strongbox-cron/strongbox-cron-tasks/pom.xml:169]: 1. Cosmetic: This is not properly indented.
2. The version for this should only be defined in the `strongbox-parent`.

File: strongbox-security/strongbox-authentication-registry/pom.xml [hunk 2]
```diff
@@ -81,14 +81,8 @@
...
-            <artifactId>apacheds-core</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
-            <groupId>org.apache.directory.server</groupId>
-            <artifactId>apacheds-protocol-ldap</artifactId>
-            <scope>test</scope>
+            <groupId>com.unboundid</groupId>
+			<artifactId>unboundid-ldapsdk</artifactId>
         </dependency>
```
Comment (carlspring) [inline strongbox-security/strongbox-authentication-registry/pom.xml:86]: Cosmetic: Incorrectly indented.

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java [hunk 3]
```diff
@@ -198,7 +196,7 @@ public void ldapConfigurationTestRequiresValidUrl()
                .body("errors[0]['name']", equalTo("configuration.url"))
                .body("errors[0]['messages'][0]", equalTo("must be a valid URI"));
     }
-
+	@Disabled
```
Comment (carlspring) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java:199]: Why was this test disabled?

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java [hunk 4]
```diff
@@ -447,10 +445,11 @@ private LdapConfigurationTestForm getLdapConfigurationTestForm()
...
-        public HazelcastInstanceId hazelcastInstanceIdLacct() {
+        public HazelcastInstanceId hazelcastInstanceIdLacct()
+        {
             return new HazelcastInstanceId("LdapAuthenticatorConfigurationControllerTest-hazelcast-instance");
         }
 
     }
 
-}
+}
```
Comment (carlspring) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java:455]: Please, add a new line at the end of this file. Thanks!

File: strongbox-cron/strongbox-cron-tasks/pom.xml (hunk 1 above)
Comment (mknutsen79) [reply strongbox-cron/strongbox-cron-tasks/pom.xml]: Okay, fixed that, I thought I had removed it but apparently I hadn't.

File: strongbox-security/strongbox-authentication-registry/pom.xml (hunk 2 above)
Comment (mknutsen79) [reply strongbox-security/strongbox-authentication-registry/pom.xml:86]: I am unsure why it looks like this, it looks fine on my end but I tried to fix it.

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java (hunk 3 above)
Comment (mknutsen79) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java:199]: A problem occurs with this test file and to keep it in line with the backport the IOException was removed and disabled was added.

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java (hunk 4 above)
Comment (mknutsen79) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java]: Done!

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java (hunk 3 above)
Comment (carlspring) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java:199]: Did I break it (was it commented out in my pull), or does it only fail for you?
You would have to provide some more details regarding "what fails".

Comment (mknutsen79) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java:199]: The error ran into was:
Tests run: 15, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 29.202 s <<< FAILURE! - in org.carlspring.strongbox.controllers.configuration.security.ldap.LdapAuthenticatorConfigurationControllerTest
ldapConfigurationTestShouldFailWithInvalidConfiguration  Time elapsed: 0.339 s  <<< FAILURE!
java.lang.AssertionError: 
1 expectation failed.
Expected status code <200> but was <500>.

	at org.carlspring.strongbox.controllers.configuration.security.ldap.LdapAuthenticatorConfigurationControllerTest.ldapConfigurationTestShouldFailWithInvalidConfiguration(LdapAuthenticatorConfigurationControllerTest.java:225)

Then I found I didn't have this class set up in PR-1783. The changes made here are to match that PR.

File: strongbox-cron/strongbox-cron-tasks/pom.xml [hunk 5]
```diff
@@ -163,12 +163,8 @@
...
 
         <dependency>
-            <groupId>org.apache.directory.server</groupId>
-            <artifactId>apacheds-core</artifactId>
-        </dependency>
-        <dependency>
-            <groupId>org.apache.directory.server</groupId>
-            <artifactId>apacheds-protocol-ldap</artifactId>
+            <groupId>com.unboundid</groupId>
+			<artifactId>unboundid-ldapsdk</artifactId>
```
Comment (carlspring) [inline strongbox-cron/strongbox-cron-tasks/pom.xml:167]:
```suggestion
            <groupId>com.unboundid</groupId>
            <artifactId>unboundid-ldapsdk</artifactId>
```

File: strongbox-security/strongbox-authentication-registry/pom.xml [hunk 6]
```diff
@@ -81,14 +81,8 @@
...
-            <groupId>org.apache.directory.server</groupId>
-            <artifactId>apacheds-core</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
-            <groupId>org.apache.directory.server</groupId>
-            <artifactId>apacheds-protocol-ldap</artifactId>
-            <scope>test</scope>
+            <groupId>com.unboundid</groupId>
+			<artifactId>unboundid-ldapsdk</artifactId>
```
Comment (carlspring) [inline strongbox-security/strongbox-authentication-registry/pom.xml:85]:
```suggestion
            <groupId>com.unboundid</groupId>
            <artifactId>unboundid-ldapsdk</artifactId>
```

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java (hunk 3 above)
Comment (carlspring) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java:199]: Is that the whole stacktrace of the exception?
Does this fail in your IDE, or in your terminal, or both?

Comment (mknutsen79) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java:199]: It failed in the terminal. Once I made the change to match the change made to it in PR-1783 everything worked fine.

Comment (carlspring) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java:199]: Does this mean you've removed the `@Disabled` part and re-enabled the test? This test needs to be enabled and looked into.

@madhukarbharti / @luan-cestari : Would you like to help @mknutsen79 investigate this failing test?

Comment (mknutsen79) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java:199]: I haven't gotten to removing the disabled and testing the test yet, been having to do some other things. The plan is to get on that this coming week if i'm not beaten to it yet.

Comment (carlspring) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java]: Any update on this? :)

Comment (mknutsen79) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java]: Yes, the latest commit should have the change made to the test. I think I found what the issue was in the test and made the change.

Comment (carlspring) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java]: Great! Did you push the code, after you re-enabled the test?

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java (hunk 4 above)
Comment (carlspring) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java]: Thanks! ;)

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java (hunk 3 above)
Comment (mknutsen79) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java]: Yes I did, it is in this latest pull request that I did on monday.

Comment (carlspring) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java]: I'm sorry... which one is that and do we then need to close this one instead?

Comment (mknutsen79) [reply strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/security/ldap/LdapAuthenticatorConfigurationControllerTest.java]: It should still be part of this request, its the latest commit i've done on this branch. Sorry for the bad wording there.

**Review by carlspring — CHANGES_REQUESTED** [review]
I have reviewed your fixed and, while they do look good, you will need to rebase (please check [our article](https://strongbox.github.io/developer-guide/git/rebase-vs-merge.html) on this topic), because there are recent changes from the upstream that are breaking your tests. Once you have rebased, these should be resolved and the build on our CI server should pass.

**Comment by carlspring** [comment]
Thank you very much for backporting these fixes, @mknutsen79 ! :)

The [Jsnkins build for this](https://jenkins.carlspring.org/blue/organizations/jenkins/strongbox%2Fbuilds%2Fstrongbox/detail/PR-1908/10/pipeline/ has now passed successfully. There are now two remaining things to do:
* [ ] Switch the version of the `strongbox-parent` to `1.0-SNAPSHOT` (in the `strongbox-parent`).
* [ ] Switch the version of the `strongbox-parent` to `1.0-SNAPSHOT` (in the `strongbox`).

Could you please look into this, so we could finalize this task? Thanks! :)

**Review by carlspring — APPROVED** [review]
Looks good to me!

Thank you for your excellent help with getting this backported from our other branch! :)

**Pull request merged**